contracts/exchange-forwarder/contracts/src/MixinExchangeWrapper.sol

```diff
@@ -77,183 +80,243 @@ contract MixinExchangeWrapper is
-
-            // We assume that asset being bought by taker is the same for each order.
-            // We assume that asset being sold by taker is WETH for each order.
-            orders[i].makerAssetData = makerAssetData;
-            orders[i].takerAssetData = wethAssetData;
-
-            // Calculate the remaining amount of makerAsset to buy
-            uint256 remainingMakerAssetFillAmount = _safeSub(makerAssetFillAmount, totalFillResults.makerAssetFilledAmount);
-
-            // Convert the remaining amount of makerAsset to buy into remaining amount
-            // of takerAsset to sell, assuming entire amount can be sold in the current order.
-            // We round up because the exchange rate computed by fillOrder rounds in favor
-            // of the Maker. In this case we want to overestimate the amount of takerAsset.
-            uint256 remainingTakerAssetFillAmount = _getPartialAmountCeil(
-                orders[i].takerAssetAmount,
-                orders[i].makerAssetAmount,
-                remainingMakerAssetFillAmount
+        // Percentage fee
+        if (order.takerFeeAssetData.equals(order.makerAssetData)) {
+            // Calculate the remaining amount of takerAsset to sell
+            uint256 remainingTakerAssetFillAmount = LibMath.getPartialAmountCeil(
+                order.takerAssetAmount,
+                order.makerAssetAmount.safeSub(order.takerFee),
+                makerAssetBuyAmount.safeSub(makerAssetAcquiredAmount)
             );
 
             // Attempt to sell the remaining amount of takerAsset
-            FillResults memory singleFillResults = _fillOrderNoThrow(
-                orders[i],
+            singleFillResults = _fillOrderNoThrow(
+                order,
                 remainingTakerAssetFillAmount,
-                signatures[i]
+                signature
+            );
+        // WETH fee
+        } else if (order.takerFeeAssetData.equals(order.takerAssetData)) {
+            // Calculate the remaining amount of takerAsset to sell
+            uint256 remainingTakerAssetFillAmount = LibMath.getPartialAmountCeil(
+                order.takerAssetAmount,
+                order.makerAssetAmount,
+                makerAssetBuyAmount.safeSub(makerAssetAcquiredAmount)
             );
 
-            // Update amounts filled and fees paid by maker and taker
-            _addFillResults(totalFillResults, singleFillResults);
-
-            // Stop execution if the entire amount of makerAsset has been bought
-            makerAssetFilledAmount = totalFillResults.makerAssetFilledAmount;
-            if (makerAssetFilledAmount >= makerAssetFillAmount) {
-                break;
-            }
+            // Attempt to sell the remaining amount of takerAsset
+            singleFillResults = _fillOrderNoThrow(
+                order,
+                remainingTakerAssetFillAmount,
+                signature
+            );
+        } else {
```

Comment:
Thinking about it more, should there be a third case where `takerFee == 0`? It is possible that an order has null or random `takerFeeAssetData` with a 0 fee, which should still be supported by this contract. That should probably be the first case we check for both `_marketBuySingleOrder` and `_marketSellSingleOrder` since it is the simplest.